File: sdk/python/src/tinyhumans/api/mascots.py
from __future__ import annotations

from typing import Any

from ..http import Json
from ._base import ApiNamespace, enc

__all__ = ["MascotsApi"]


class MascotsApi(ApiNamespace):
    """Mascots: library, manifests, Rive assets, the interactive demo, and meeting dispatch."""

    def list_mascots(self, **kwargs: Any) -> Json:
        """List available mascots."""
        return self._http.get("/mascots", **kwargs)

    def get_demo(self, **kwargs: Any) -> Json:
        """Interactive mascot library demo page."""
        return self._http.get("/mascots/demo", **kwargs)

    def join_meeting(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Send the mascot bot into a live meeting."""
        return self._http.post("/mascots/join-meeting", body=body, **kwargs)

    def list_meetings(
        self, query: dict[str, Any] | None = None, **kwargs: Any
    ) -> Json:
        """List the authenticated user's mascot-bot meeting history."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get("/mascots/meetings", query=merged, **kwargs)

    def get_mascot(self, id: str, **kwargs: Any) -> Json:
        """Get a mascot manifest."""
        return self._http.get(f"/mascots/{enc(id)}", **kwargs)

    def get_mascot_riv(self, id: str, **kwargs: Any) -> Json:
        """Download the Rive animation file for a mascot."""
        return self._http.get(f"/mascots/{enc(id)}/riv", **kwargs)

File: src/api/mascots.rs
//! Mascots: library, manifests, Rive assets, the interactive demo, and meeting dispatch.

use reqwest::Method;
use serde_json::Value;

use super::types::JoinMeetingRequest;
use crate::{enc, Error, HttpClient, QueryParam};

/// Typed client for the `/mascots/*` routes.
pub struct MascotsApi<'a> {
    http: &'a HttpClient,
}

impl<'a> MascotsApi<'a> {
    pub fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    /// List available mascots.
    pub async fn list_mascots(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/mascots", &[], None, true)
            .await
    }

    /// Interactive mascot library demo page.
    pub async fn get_demo(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/mascots/demo", &[], None, true)
            .await
    }

    /// Send the mascot bot into a live meeting.
    pub async fn join_meeting(&self, request: &JoinMeetingRequest) -> Result<Value, Error> {
        let body = serde_json::to_value(request).expect("join meeting request is serializable");
        self.http
            .send(
                Method::POST,
                "/mascots/join-meeting",
                &[],
                Some(&body),
                true,
            )
            .await
    }

    /// List the authenticated user's mascot-bot meeting history.
    pub async fn list_meetings(&self, query: &[QueryParam]) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/mascots/meetings", query, None, true)
            .await
    }

    /// Get a mascot manifest.
    pub async fn get_mascot(&self, id: &str) -> Result<Value, Error> {
        let path = format!("/mascots/{}", enc(id));
        self.http.send(Method::GET, &path, &[], None, true).await
    }

    /// Download the Rive animation file for a mascot.
    pub async fn get_mascot_riv(&self, id: &str) -> Result<Value, Error> {
        let path = format!("/mascots/{}/riv", enc(id));
        self.http.send(Method::GET, &path, &[], None, true).await
    }
}
